rllib/models/base_model.py

```diff
@@ -0,0 +1,322 @@
+                state to be passed as the first state of the next rollout.
+
+        Examples:
+            >>> output, state = model.unroll(TensorDict(...), TensorDict(...))
+            >>> output # TensorDict(...)
+            >>> state # TensorDict(...)
+
+        """
+        self.input_spec.validate(inputs)
+        self.prev_state_spec.validate(prev_state)
+        # We hide inputs not specified in input_spec to prevent accidental use.
+        inputs = inputs.filter(self.input_spec)
+        prev_state = prev_state.filter(self.prev_state_spec)
+        inputs, prev_state = self._check_inputs_and_prev_state(inputs, prev_state)
+        outputs, next_state = self._unroll(inputs, prev_state, **kwargs)
+        self.output_spec.validate(outputs)
+        self.next_state_spec.validate(next_state)
+        outputs, next_state = self._check_outputs_and_next_state(outputs, next_state)
+        return outputs, next_state
+
+    def _check_inputs_and_prev_state(
+        self, inputs: TensorDict, prev_state: TensorDict
+    ) -> Tuple[TensorDict, TensorDict]:
+        """Override this function to add additional checks on inputs.
+
+        Args:
+            inputs: TensorDict containing inputs to the model
+            prev_state: The previous recurrent state
+
+        Returns:
+            inputs: Potentially modified inputs
+            prev_state: Potentially modified recurrent state
+        """
+        return inputs, prev_state
+
+    def _check_outputs_and_next_state(
+        self, outputs: TensorDict, next_state: TensorDict
+    ) -> Tuple[TensorDict, TensorDict]:
+        """Override this function to add additional checks on outputs.
+
+        Args:
+            outputs: TensorDict output by the model
+            next_state: Recurrent state output by the model
+
+        Returns:
+            outputs: Potentially modified TensorDict output by the model
+            next_state: Potentially modified recurrent state output by the model
+        """
+        return outputs, next_state
+
+
+class Model(RecurrentModel):
+    """A RecurrentModel made non-recurrent by ignoring
+    the input/output states.
+
+    As a convienience, users may override _forward instead of _unroll,
+    which hides model states.
+
+    Args:
+        name: An optional name for the module
+    """
+
+    @property
+    @override(RecurrentModel)
+    def prev_state_spec(self) -> SpecDict:
+        return SpecDict()
+
+    @property
+    @override(RecurrentModel)
+    def next_state_spec(self) -> SpecDict:
+        return SpecDict()
+
+    @override(RecurrentModel)
+    def _initial_state(self) -> TensorDict:
+        return TensorDict()
+
+    @override(RecurrentModel)
+    def _check_inputs_and_prev_state(
+        self, inputs: TensorDict, prev_state: TensorDict
+    ) -> Tuple[TensorDict, TensorDict]:
+        inputs = self._check_inputs(inputs)
+        return inputs, prev_state
+
+    def _check_inputs(self, inputs: TensorDict) -> TensorDict:
```

Comment:
can these _check_xxx APIs return True/False. think of them taking const inputs in C++ world.
I for some reasons really hate these validation methods make modifications to the data to be checked.
there was one time I got super confused because our Env checking logics actually creates additional data fields on my env.

Reply:
I agree with @gjoliver with this one. Let's rename them to `_update_xxx_before_fwd` and `_update_xxx_after_fwd` to be super explicit.

Reply:
fixed.